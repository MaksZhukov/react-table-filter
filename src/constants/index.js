const SUB_ACTIONS_TYPES = ['SUCCESS', 'ERROR', 'PENDING'];
const TYPE_SEARCH = {
  all: '**',
  startWith: '*_',
  exact: '""',
};
const ORDER_CELLS = {
  up: 'A-Z',
  down: 'Z-A',
};
const OFFSETS_FILTERS = {
  height: 300,
  width: 400,
  indentX: 110,
  indentY: 100,
  additionalIndentY: 10,

};
export {
  SUB_ACTIONS_TYPES,
  TYPE_SEARCH,
  ORDER_CELLS,
  OFFSETS_FILTERS,
};
