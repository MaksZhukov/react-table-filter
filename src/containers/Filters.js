import {
  connect,
} from 'react-redux';
import Filter from '../components/Filters';
import {
  getTables, changeContexts, changeDimensions,
  changeCells, changeCellsAll, inputSearch, changeOrderCells,
  changeSearchType,
} from '../actions';


const mapStateToProps = (state, { id }) => ({
  filters: state.get('panels').find(panel => panel.get('id') === id).get('filters'),
  responseGetTables: state.get('panels').find(panel => panel.get('id') === id).get('responseGetTables'),
});

const mapDispatchToProps = dispatch => ({
  getTables: id => dispatch(getTables(id)),
  changeContexts: ({ id, contexts }) => dispatch(changeContexts({ id, contexts })),
  changeDimensions: ({ id, dimensions }) => dispatch(changeDimensions({ id, dimensions })),
  changeCells: ({ id, cell }) => dispatch(changeCells({ id, cell })),
  changeCellsAll: id => dispatch(changeCellsAll(id)),
  inputSearch: ({ id, value }) => dispatch(inputSearch({ id, value })),
  changeSearchType: ({ id, type }) => dispatch(changeSearchType({ id, type })),
  changeOrderCells: id => dispatch(changeOrderCells(id)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Filter);
