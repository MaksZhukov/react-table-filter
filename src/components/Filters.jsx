import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import Multiselect from 'react-widgets/lib/Multiselect';
import DropdownList from 'react-widgets/lib/DropdownList';
import SearchInput from 'react-search-input';
import Checkbox from 'rc-checkbox';
import { Map } from 'immutable';
import { TYPE_SEARCH } from '../constants';
import 'rc-checkbox/assets/index.css';

class Filters extends PureComponent {
  componentDidMount() {
    const { getTables, id, responseGetTables } = this.props;
    if (!responseGetTables.status || responseGetTables.status !== 'success') {
      getTables(id);
    }
  }

  handleChangeContext = (values) => {
    const { changeContexts, id } = this.props;
    changeContexts({ id, contexts: values });
  }

  handleChangeDimensions = (values) => {
    const { changeDimensions, id } = this.props;
    changeDimensions({ id, dimensions: values });
  }

  handleChangeCell = ({ target }) => {
    const { changeCells, id } = this.props;
    changeCells({ id, cell: target.value });
  }

  handleChangeCellAll = () => {
    const { changeCellsAll, id } = this.props;
    changeCellsAll(id);
  }

  handleInputSearch = ({ target }) => {
    const { inputSearch, id } = this.props;
    inputSearch({ id, value: target.value });
  }

  stopDraggableFilters = (event) => {
    event.stopPropagation();
  }

  handleChangeSearchType = (value) => {
    const { changeSearchType, id } = this.props;
    changeSearchType({ id, type: value });
  }

  handleChangeOrderCells = () => {
    const { changeOrderCells, id } = this.props;
    changeOrderCells(id);
  }

  render() {
    const { filters, id } = this.props;
    return (
      <>
        <div className="filters-header">
          Filters for â„–
          {' '}
          {id}
        </div>
        <div className="filters-body">
          CONTEXTS
          <div className="filters-context">
            <Multiselect
              value={filters.get('contexts').toJS()}
              data={filters.get('defaultContexts').toJS()}
              onChange={this.handleChangeContext}
            />
          </div>
          DIMENSIONS
          <div className="filters-dimensions">
            <Multiselect
              value={filters.get('dimensions').toJS()}
              data={filters.get('defaultDimensions').toJS()}
              onChange={this.handleChangeDimensions}
            />
          </div>
          SEARCH
          <div className="filters-search">
            <SearchInput onMouseDown={this.stopDraggableFilters} onInput={this.handleInputSearch} value={filters.getIn(['search', 'value'])} className="search-input" />
            <div className="filters-search-type">
              <DropdownList
                defaultValue={filters.getIn(['search', 'type'])}
                data={[
                  TYPE_SEARCH.all,
                  TYPE_SEARCH.startWith,
                  TYPE_SEARCH.exact,
                ]}
                onChange={this.handleChangeSearchType}
              />
            </div>
            <button type="button" onClick={this.handleChangeOrderCells}>
              {filters.get('orderCells')}
            </button>
          </div>
          CELLS
          <div className="filters-cells">
            {filters.get('cells').filter(cell => cell.get('visible')).size > 0
              && (
              <div className="filters-cells-all">
                <label>
                  <Checkbox checked={filters.get('checkedAllCells')} onChange={this.handleChangeCellAll} />
                  &nbsp; All
                </label>
              </div>
              )
            }
            {filters.get('cells').filter(cell => cell.get('visible')).map(cell => (
              <div key={cell} className="filters-cells-cell">
                <label>
                  <Checkbox value={cell} checked={cell.get('checked')} onChange={this.handleChangeCell} />
                  &nbsp;
                  {cell.get('value')}
                </label>
              </div>
            ))}
          </div>
        </div>
      </>
    );
  }
}

Filters.propTypes = {
  id: PropTypes.number.isRequired,
  filters: PropTypes.instanceOf(Map).isRequired,
  responseGetTables: PropTypes.instanceOf(Map).isRequired,
  getTables: PropTypes.func.isRequired,
  changeContexts: PropTypes.func.isRequired,
  changeDimensions: PropTypes.func.isRequired,
  changeCells: PropTypes.func.isRequired,
  changeCellsAll: PropTypes.func.isRequired,
  inputSearch: PropTypes.func.isRequired,
  changeSearchType: PropTypes.func.isRequired,
  changeOrderCells: PropTypes.func.isRequired,
};

export default Filters;
