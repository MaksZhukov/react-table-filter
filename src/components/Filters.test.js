import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { shallowToJson } from 'enzyme-to-json';
import { fromJS, Map } from 'immutable';
import Filters from './Filters';
import { TYPE_SEARCH, ORDER_CELLS } from '../constants';


configure({ adapter: new Adapter() });
describe('component Filters', () => {
  const filters = fromJS({
    defaultContexts: [],
    contexts: [],
    defaultDimensions: [],
    dimensions: [],
    cells: [],
    checkedAllCells: false,
    orderCells: ORDER_CELLS.up,
    search: {
      value: '',
      type: TYPE_SEARCH.all,
    },
  });
  const props = {
    id: 0,
    filters,
    responseGetTables: Map({}),
    getTables: jest.fn(),
    changeContexts: jest.fn(),
    changeDimensions: jest.fn(),
    changeCells: jest.fn(),
    changeCellsAll: jest.fn(),
    inputSearch: jest.fn(),
    changeSearchType: jest.fn(),
    changeOrderCells: jest.fn(),
  };
  it('render component Filters without crashing', () => {
    const wrapper = shallow(
      <Filters {...props} />,
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
  it('render component Filters no have state', () => {
    const wrapper = shallow(
      <Filters {...props} />,
    );
    expect(wrapper.instance().state).toBe(null);
  });

  it('render component Filters has search component', () => {
    const wrapper = shallow(
      <Filters {...props} />,
    );
    expect(wrapper.find('Search').length).toBe(1);
  });
  it('render component Filters event input search works properly', () => {
    const expectData = { id: 5, value: 4 };
    let inputData = null;
    const wrapper = shallow(
      <Filters {...{
        ...props,
        ...{ id: 5, inputSearch: jest.fn((data) => { inputData = data; }) },
      }}
      />,
    );
    wrapper.find('Search').simulate('input', { target: { value: 4 } });
    expect(expectData).toEqual(inputData);
  });
  it('render component Filters contain text CELLS, DIMENSIONS, CONTEXTS, SEARCH', () => {
    const wrapper = shallow(<Filters {...props} />);
    expect(wrapper.contains('SEARCH')).toBe(true);
    expect(wrapper.contains('DIMENSIONS')).toBe(true);
    expect(wrapper.contains('CONTEXTS')).toBe(true);
    expect(wrapper.contains('CELLS')).toBe(true);
  });
});
