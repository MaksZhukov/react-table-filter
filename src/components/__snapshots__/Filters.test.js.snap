// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component Filters render component Filters without crashing 1`] = `
<Fragment>
  <div
    className="filters-header"
  >
    Filters for â„–
     
    0
  </div>
  <div
    className="filters-body"
  >
    CONTEXTS
    <div
      className="filters-context"
    >
      <Uncontrolled(Multiselect)
        data={Array []}
        onChange={[Function]}
        value={Array []}
      />
    </div>
    DIMENSIONS
    <div
      className="filters-dimensions"
    >
      <Uncontrolled(Multiselect)
        data={Array []}
        onChange={[Function]}
        value={Array []}
      />
    </div>
    SEARCH
    <div
      className="filters-search"
    >
      <Search
        caseSensitive={false}
        className="search-input"
        fuzzy={false}
        onChange={[Function]}
        onInput={[Function]}
        onMouseDown={[Function]}
        throttle={200}
        value=""
      />
      <div
        className="filters-search-type"
      >
        <Uncontrolled(DropdownList)
          data={
            Array [
              "**",
              "*_",
              "\\"\\"",
            ]
          }
          defaultValue="**"
          onChange={[Function]}
        />
      </div>
      <button
        onClick={[Function]}
        type="button"
      >
        A-Z
      </button>
    </div>
    CELLS
    <div
      className="filters-cells"
    />
  </div>
</Fragment>
`;
